### ISSUE TITLE
Interval(x, x) vs. Interval(oo, x) inconsistency?

### ISSUE THREAD

**Issue by gschintgen** [sympy/sympy#17488]
There seems to be some minor inconsistency with `Interval`'s handling of general symbols:
```
In [63]: Interval(x,x)
Out[63]: {x}

In [64]: Interval(oo,oo)
Out[64]: ∅

In [65]: Interval(oo, x)
Out[65]: [∞, x]

In [66]: r = Symbol('r', real=True)

In [67]: Interval(oo, r)
Out[67]: ∅
```
In `Out[65]` the given interval isn't simplified to `EmptySet`, presumably because `x` may be non-real (compare with `Out[67]`), but then it shouldn't evaluate `Interval(x, x)` either. I'm just not sure which one should be changed, 63 or 65.

Is there a consensus on what SymPy is supposed to do in situations where only reals are meaningful, but a general symbol is passed? Assume it to be real or have the user / library code explicitly specify the realness?

**Comment by asmeurer** [comment]
A non-real to Interval isn't meaningful, since Interval(a, b) means the numbers x such that a < x and x < b. If anything it should produce an empty set, since x < b would be false for nonreal x. Presumably 65 not simplifying is just a bug in the assumptions testing. 

Also note that infinite end points in Interval always imply open, because Interval is always a subset of Reals. So 65 giving EmptySet but 63 not is consistent. 

> Is there a consensus on what SymPy is supposed to do in situations where only reals are meaningful, but a general symbol is passed? Assume it to be real or have the user / library code explicitly specify the realness?

The common thing is to implicitly assume a necessary assumption, failing if that assumption is False but not failing if it is None. That way people don't have to manually set assumptions for things to work.

**Comment by gschintgen** [comment]
> If anything it should produce an empty set, since x < b would be false for nonreal x. Presumably 65 not simplifying is just a bug in the assumptions testing.
> [...]
> The common thing is to implicitly assume a necessary assumption, failing if that assumption is False but not failing if it is None. That way people don't have to manually set assumptions for things to work.

Thanks for the quick and detailed response, in particular the part about implicit necessary assumptions. So 63 is indeed fine, since the necessary assumption of x being real is implicitly assumed to hold. But 65 should ideally auto-evaluate to EmptySet. Got it.

**Comment by Soniyanayak51** [comment]
@asmeurer all intervals starting with oo should retuen an empty set, but I'm not sure about Interval.(oo, oo).
Should it return oo or an empty set?

**Comment by gschintgen** [comment]
An empty set, since oo is not a real number, so FiniteSet(oo) would be non-sensical in this case.

**Comment by jksuom** [comment]
> Interval is always a subset of Reals

I have always wondered what was the reason for this. The set of extended reals is an ordered set where intervals, open or closed, with any two enpoints, including `oo` and `-oo` can be naturally defined.

**Comment by gschintgen** [comment]
@jksuom:
> > Interval is always a subset of Reals
> 
> I have always wondered what was the reason for this. The set of extended reals is an ordered set where intervals, open or closed, with any two enpoints, including `oo` and `-oo` can be naturally defined.

Good point! Especially given that the assumption system(s) also use the extended reals.

(The defaults should stay unchanged though: infinite endpoints will be excluded unless the user explicitly specifies e.g. `right_open = False`)

**Comment by asmeurer** [comment]
> I have always wondered what was the reason for this. The set of extended reals is an ordered set where intervals, open or closed, with any two enpoints, including oo and -oo can be naturally defined.

There may be a good reason for it, though I can't remember any other than purity reasons. Maybe there are some calculations that are easier if we can always assume Interval is a subset of Reals? 

There is also currently this behavior:

```py
>>> Interval(0, x).subs(x, oo)
Interval(0, oo)
```

which I think is wrong. `x` is in `Interval(0, x)`, because it get created with closed end-points, but `oo` is not in the subs'ed interval. OTOH subs doesn't always produce a strictly mathematically equivalent object, so perhaps this can be chalked up to that. 

Would any calculations be simpler if we allowed infinities in closed intervals? 

One thing I will say is I think that if you have something like [x, y] as above, then x and y should be considered implicitly finite. So the above subs is somewhat of a garbage in-garbage out scenario, because you're replacing a symbol that cannot be infinity with infinity.
